### DISCUSSION TITLE
Code related : How to fill the sharp corner?

### DISCUSSION THREAD

**Discussion by 601469588** [inkstitch/inkstitch#1603] [Q&A]
I've noticed that many people are asking questions about software operation. But my question is about algorithm implementation.
Recently, I am studying the problems related to automatic embroidery. I found some papers on the generation method of Satin Stitch, like '**Automatic embroidery texture synthesis for garment design and online display**'. I noted that both theirs' method and your software is to segment the shapes, and then filling them.
But there's one thing I can't figure out about filling. How to fill the sharp corner?(Like the Issue: [https://github.com/inkstitch/inkstitch/issues/308]). On the inside of a sharp corner, how does the code delete the stitches? And on the outside of a sharp corner, how does the code add the stitches? May be you can tell me the code location of the above processing, and I'll study it myself.
Thanks a lot.

Update : 2022.03.10 ----------------------------------------------
I spent times to retrieve related questions. I find that you have discussed this problem.(#214, #33) And I read the code in auto_fill.py. I still did not understand how it work on the sharp corner. I know you segmenting the shape with parallel line, but how it works on the inside and outside of a sharp corner?
As shown below, I use tick red parallel lines to segment the shape like you said. I wonder how the parallel lines are tiled in the sharp corner? How does the code described or handled the inside(green circle) and outside(yellow circle) of the corner?
![corner](https://user-images.githubusercontent.com/87968109/157627923-c0f2747e-1685-4968-a90d-11c3ac436468.png)

**Comment by 601469588** [comment]
One-to-many mapping? Triangulation? Or some mathematical way?

**Reply by lexelby** [reply]
How does the code handle it? Poorly.

First, you'll want to be looking at lib/elements/satin_column.py for this code.  This is all about satins, and auto_fill.py is about fill stitch, which is an entirely different thing.

Anyway, on to your questions.  As far as the inside of a sharp corner, we have not yet implemented the relocation of stitches to avoid piling up.  I'm not yet sure how to do it.  The best guess I have would be to keep track of a running average of distance between successive stitches, and if it's too close, pull the stitch in toward the center of the satin by some amount.  I'm not sure how well it will work, though.

We've tried several times to tackle the problem of thinned stitches on the outside of a sharp corner.  It's _way_ better than it was in the past, but it's still pretty bad.  There are a lot of comments in satin_column.py about how I tried to compensate for the thinning effect.  I'm afraid I can't remember the name of the method in question right now, but it should be fairly clear.  I'd be happy to answer any questions about how it works.

**Reply by 601469588** [reply]
Thank you for your answer.
I think I made a mistake at first. I thought your processing was based on bitmap which only have information about edges and interior points. So I'm confused how you can divide them into small segments so perfectly. Now, I found out that you handled it based on SVG.
I'm not very familiar with SVG. Here's my guess. I don't know if it's right. _PATH_ is a parameter of SVG, like a skeleton of the bitmap shape. It can represent the shape of an entire SVG graph. Your segmentation algorithm is based on _PATH_. 
Although the segmentation shapes have overlapping and intersecting. But you regard them as railway tracks, and only extract the two parts of the intersection as the basis for segmentation. Of course, there are some special cases to deal with.
Then you traverse them in a way similar to a bipartite graph.
Do I understand correctly?

**Reply by lexelby** [reply]
I'm not quite sure I follow your explanation, but I think it's about right.  SVG defines paths exactly using polynomials (bezier curves).  Ink/Stitch approximates that into a sequence of contiguous line segments -- called a "LineString" in some mathematical circles.  The Satin Column code walks down the two rails, trying to step an even distance down each based on the "train tracks" that the user has placed.

It's a flexible method, but it does run into problems.  I've been toying with the idea of coming up with a new approach for awhile, but I haven't figured out what that might be.
